File: src/main/java/co/uk/isxander/evergreenhud/settings/impl/BooleanSetting.java
/*
 * Copyright (C) isXander [2019 - 2021]
 * This program comes with ABSOLUTELY NO WARRANTY
 * This is free software, and you are welcome to redistribute it
 * under the certain conditions that can be found here
 * https://www.gnu.org/licenses/gpl-3.0.en.html
 *
 * If you have any questions or concerns, please create
 * an issue on the github page that can be found here
 * https://github.com/isXander/EvergreenHUD
 *
 * If you have a private concern, please contact
 * isXander @ business.isxander@gmail.com
 */

package co.uk.isxander.evergreenhud.settings.impl;

import co.uk.isxander.evergreenhud.settings.Setting;

public class BooleanSetting extends Setting {

    private final boolean def;
    private boolean value;

    public BooleanSetting(String name, String category, String description, boolean def, boolean internal) {
        super(name, description, category, internal);
        this.value = this.def = def;
    }

    public BooleanSetting(String name, String category, boolean def, boolean internal) {
        super(name, "", category, internal);
        this.value = this.def = def;
    }

    public BooleanSetting(String name, String category, String description, boolean def) {
        super(name, description ,category);
        this.value = this.def = def;
    }

    public BooleanSetting(String name, String category, boolean def) {
        super(name, category);
        this.value = this.def = def;
    }

    public boolean get() {
        return value;
    }

    public boolean getDefault() {
        return def;
    }

    public void set(boolean newVal) {
        if (onChange(value, newVal))
            value = newVal;
    }

    public boolean toggle() {
        boolean newVal = !value;
        if (onChange(value, newVal)) {
            value = newVal;
        }
        return value;
    }

    protected boolean onChange(boolean oldValue, boolean newValue) {
        return !isDisabled();
    }

    @Override
    public void reset() {
        this.value = def;
    }
}

File: core/src/main/kotlin/dev/isxander/evergreenhud/settings/impl/BooleanSetting.kt
/*
 | EvergreenHUD - A mod to improve on your heads-up-display.
 | Copyright (C) isXander [2019 - 2021]
 |
 | This program comes with ABSOLUTELY NO WARRANTY
 | This is free software, and you are welcome to redistribute it
 | under the certain conditions that can be found here
 | https://www.gnu.org/licenses/lgpl-3.0.en.html
 |
 | If you have any questions or concerns, please create
 | an issue on the github page that can be found here
 | https://github.com/isXander/EvergreenHUD
 |
 | If you have a private concern, please contact
 | isXander @ business.isxander@gmail.com
 */

package dev.isxander.evergreenhud.settings.impl

import dev.isxander.evergreenhud.settings.DataType
import dev.isxander.evergreenhud.settings.Setting
import dev.isxander.evergreenhud.settings.providers.IValueProvider
import gg.essential.elementa.UIComponent
import kotlin.reflect.KProperty1

@Target(AnnotationTarget.FIELD, AnnotationTarget.PROPERTY, AnnotationTarget.VALUE_PARAMETER)
@Retention(AnnotationRetention.RUNTIME)
@MustBeDocumented
annotation class BooleanSetting(val name: String, val category: Array<String>, val description: String, val save: Boolean = true)

class BooleanSettingWrapped(annotation: BooleanSetting, provider: IValueProvider<Boolean>) : Setting<Boolean, BooleanSetting>(annotation, provider, DataType.BOOLEAN) {
    override val name: String = annotation.name
    override val category: Array<String> = annotation.category
    override val description: String = annotation.description
    override val shouldSave: Boolean = annotation.save

    override var serializedValue: Any
        get() = value
        set(new) {
            value = new as Boolean
        }

    override val defaultSerializedValue: Any = default

    override fun addComponentToUI(parent: UIComponent) {
        TODO("Not yet implemented")
    }
}

